scripts/verify-snapshot-source-map.js

```diff
@@ -0,0 +1,443 @@
+        let inIndex = inJs.indexOf(`"${id}"`)
+        if (inIndex < 0) inIndex = inJs.indexOf(`'${id}'`)
+        if (inIndex < 0) throw new Error(`Failed to find "${id}" in input`)
+        const inLines = inJs.slice(0, inIndex).split('\n')
+        const inLine = inLines.length
+        const inColumn = inLines[inLines.length - 1].length
+
+        const expected = JSON.stringify({ source, line: inLine, column: inColumn })
+        const observed = JSON.stringify({ source, line, column })
+        recordCheck(expected === observed, `expected original position: ${expected}, observed original position: ${observed}`)
+
+        // Also check the reverse mapping
+        const positions = map.allGeneratedPositionsFor({ source, line: inLine, column: inColumn })
+        recordCheck(positions.length > 0, `expected generated positions: 1, observed generated positions: ${positions.length}`)
+        let found = false
+        for (const { line, column } of positions) {
+          if (line === outLine && column === outColumn) {
+            found = true
+            break
+          }
+        }
+        const expectedPosition = JSON.stringify({ line: outLine, column: outColumn })
+        const observedPositions = JSON.stringify(positions)
+        recordCheck(found, `expected generated position: ${expectedPosition}, observed generated positions: ${observedPositions}`)
+      }
+    }
+
+    const sources = JSON.parse(outJsMap).sources
+    for (let source of sources) {
+      if (sources.filter(s => s === source).length > 1) {
+        throw new Error(`Duplicate source ${JSON.stringify(source)} found in source map`)
+      }
+    }
+
+    const outMap = await new SourceMapConsumer(outJsMap)
+    checkMap(outJs, outMap)
+
+    // Check that every generated location has an associated original position.
+    const outLines = outJs.trimRight().split('\n');
+
+    let insideGeneratedCode = false
+    for (let outLine = 0; outLine < outLines.length; outLine++) {
+      if (insideGeneratedCode) {
+        for (let outColumn = 0; outColumn <= outLines[outLine].length; outColumn++) {
+          const { line, column } = outMap.originalPositionFor({ line: outLine + 1, column: outColumn })
+
+          recordCheck(line !== null && column !== null, `missing location for line ${outLine+1} and column ${outColumn}`)
+        }
+      }
+
+      if (/^__commonJS\[\"\S+.js/.test(outLines[outLine])) {
+        insideGeneratedCode = true
+      }
+
+      if (outLines[outLine].startsWith('}')) {
+        insideGeneratedCode = false
+      }
+    }
+
+    if (!failed) removeRecursiveSync(tempDir)
+  }
+
+  catch (e) {
+    console.error(`❌ [${kind}] ${e && e.message || e}`)
+    failed++
+  }
+
+  return failed
+}
+
+async function main() {
+  const promises = []
+  for (const crlf of [false, true]) {
+    for (const status of ['all-healthy', 'all-deferred', 'all-norewrite']) {
+      const suffix = (crlf ? '-crlf' : '') + '-' + status
+      promises.push(
```

Comment:
Mind explaining what these test cases? they appear to be very "go" like in a JS file and I'm not sure what I should be validating is correct? What was old vs new behaviors?

Reply:
Most of these test cases are used by the existing esbuild source map verification (https://github.com/cypress-io/esbuild/blob/main/scripts/verify-source-map.js#L449-L530). I just added the ones that are relevant for snapshots as part of the snapshot source map verification as well. I added comments for the snapshot specific test cases that I added later: [`f972bb9` (#10)](https://github.com/cypress-io/esbuild/pull/10/commits/f972bb944f1be3e859325b53b7fee7b275ec5097)